packages/@jsii/go-runtime/jsii-runtime-go/internal/kernel/conversions.go

```diff
@@ -63,14 +63,19 @@ func (c *Client) castAndSetToPtr(ptr reflect.Value, data reflect.Value) {
 			return
 		}
 
+		targetType := ptr.Type()
+		if typ, ok := c.Types().FindType(ref.TypeFQN()); ok && typ.AssignableTo(ptr.Type()) {
+			targetType = typ
```

Comment:
Add a comment here for clarity on the intent

```suggestion
            // Specialize the return type to be the dynamic value type
			targetType = typ
```